cdap-api/src/main/java/co/cask/cdap/api/common/Scope.java

```diff
@@ -21,18 +21,23 @@
 import java.util.Map;
 
 /**
- * Enum for the different types of scopes. For now, only "dataset", but more will come.
+ * Enum for the different types of scopes.
  */
 public enum Scope {
-  DATASET("dataset");
+  DATASET("dataset", true),
+  MAPREDUCE("mapreduce", false),
+  SPARK("spark", false),
+  CUSTOM_ACTION("custom_action", false);
```

Comment:
I am not sure whether we need CUSTOM_ACTION. It is nice to have but I don't see the use case. For the other scopes, we are doing this to allow propagation of arguments into the scope without the need to write code. A custom action is custom code anyway and it can simply see all arguments and extract whatever it needs. If you want to keep it, I would call it WORKFLOW_ACTION or just ACTION, but not CUSTOM_ACTION.